frontend/dockerfile/parser/parser_test.go

```diff
@@ -175,3 +175,69 @@ func TestParseReturnsScannerErrors(t *testing.T) {
 	_, err := Parse(dockerfile)
 	require.EqualError(t, err, "dockerfile line greater than max allowed size of 65535")
 }
+
+func TestParseExtractsHeredoc(t *testing.T) {
+	dockerfile := bytes.NewBufferString(`
+FROM alpine:3.6
+
+RUN ls
+
+USER <<INVALID
+INVALID
+
+RUN <<EMPTY
+EMPTY
+
+RUN <<INDENT
+	foo
+	bar
+INDENT
+
+RUN <<-UNINDENT
+	baz
+	quux
+UNINDENT
+
+RUN <<-'NOEXPAND'
+	don't expand me
```

Comment:
I guess this test-case should contain an `$ENV` or `$ARG` known to the Dockerfile to verify it's not expanded by BuildKit?

Also if we have a full test, then this:

```dockerfile
ENV HELLO=bla
RUN <<-'NOEXPAND'
	HELLO=world
	echo hello $HELLO 
NOEXPAND
```

should print `hello world` (not `hello bla`)

Reply:
Have got a test for this at the parsing level, now, and will add a larger, more complete test when I get to those later on.